### PR TITLE
Bug 1003876 - Install an app from a web page

### PR DESCRIPTION
**Pull request by benfrancis**

### CODE REVIEW

**Comment by try-server-hook** [comment]
[benfrancis Ben Francis (benfrancis)](https://github.com/benfrancis) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=85da81cfac54)

**Comment by try-server-hook** [comment]
[benfrancis Ben Francis (benfrancis)](https://github.com/benfrancis) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=f224932ee736)

**Comment by try-server-hook** [comment]
[benfrancis Ben Francis (benfrancis)](https://github.com/benfrancis) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=66860b9f8164)

**Comment by try-server-hook** [comment]
[benfrancis Ben Francis (benfrancis)](https://github.com/benfrancis) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=55336bd1d49c)

**Comment by try-server-hook** [comment]
[benfrancis Ben Francis (benfrancis)](https://github.com/benfrancis) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=51391d8ff137)

File: apps/bookmark/js/web_manifest_helper.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+      };
+      xhr.send();
+    });
+  };
+
+  exports.WebManifestHelper = {
+    getManifest: getManifest
+  };
+
+})(window);
```
Comment (crdlc) [inline apps/bookmark/js/web_manifest_helper.js]: please add new line

File: apps/bookmark/js/bookmark_editor.js [hunk 2]
```diff
@@ -54,7 +70,24 @@ var BookmarkEditor = {
...
+    icon.render({'size': this.BOOKMARK_ICON_SIZE});
+  },
+  
+  _getManifest: function bookmarkEditor_getManifest(manifestURL) {
+    var manifestPromise = window.WebManifestHelper.getManifest(manifestURL);
+    
+    manifestPromise.then((function(manifestData) {
+      if (manifestData) {
+        this.appInstallationSection.classList.remove('hidden');
+        this.manifest = manifestData;
```
Comment (crdlc) [inline apps/bookmark/js/bookmark_editor.js]: We could save a reference to manifest like 

```js
var manifest = this.manifest = manifestData;
```

to use in next line although if you consider this interesting

File: apps/bookmark/js/bookmark_editor.js [hunk 3]
```diff
@@ -19,20 +21,31 @@ var BookmarkEditor = {
...
     this.cancelButton = document.getElementById('cancel-button');
     this.saveButton = document.getElementById('done-button');
+    this.appInstallationSection = document.getElementById('app-installation');
+    this.appNameText = document.getElementById('app-name');
+    this.installAppButton = document.getElementById('install-app-button');
 
     this.cancelButton.addEventListener('click', this.close.bind(this));
     this.saveListener = this.save.bind(this);
     this.saveButton.addEventListener('click', this.saveListener);
+    this.installAppButton.addEventListener('click',
```
Comment (crdlc) [inline apps/bookmark/js/bookmark_editor.js]: what do you thing about removing the listener here https://github.com/mozilla-b2g/gaia/pull/27933/files#diff-32abdeb6061956598cbede801aa9bc1aR129 ?

File: apps/bookmark/js/bookmark_editor.js [hunk 4]
```diff
@@ -19,20 +21,31 @@ var BookmarkEditor = {
...
+      this._installApp.bind(this));
 
     this.bookmarkTitle.value = this.data.name || '';
 
     this._renderIcon();
 
+    if (this.data.manifestURL) {
+      this.manifestURL = this.data.manifestURL;
+      this._getManifest(this.manifestURL);
```
Comment (crdlc) [inline apps/bookmark/js/bookmark_editor.js]: why not _setManifest instead?

File: apps/bookmark/js/bookmark_editor.js (hunk 3 above)
Comment (crdlc) [inline apps/bookmark/js/bookmark_editor.js]: Maybe I am wrong but could we move this to https://github.com/mozilla-b2g/gaia/pull/27933/files#diff-32abdeb6061956598cbede801aa9bc1aR81? or create a new method called "enableInstallSection" where we can add the listener and remote the hidden class

File: apps/bookmark/js/web_manifest_helper.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+          resolve(xhr.response);
+        } else {
+          reject(status);
+        }
+      };
+      xhr.onerror = function(e) {
+        console.error('Unable to get web manifest');
+        reject();
+      };
+      xhr.send();
```
Comment (crdlc) [inline apps/bookmark/js/web_manifest_helper.js]: reject(e.target.status) ?

File: apps/bookmark/style/save.css [hunk 6]
```diff
@@ -39,3 +38,57 @@ body[data-mode="put"] #edit-bookmark-header {
...
+  padding: 0;
+  text-align: center;
+}
+
+#app-installation #install-app-button {
+  display: block;
+  width: auto;
+  margin: 0 auto;
+  padding: 0 5rem;
+}
```
Comment (crdlc) [inline apps/bookmark/style/save.css]: add new line please

File: apps/bookmark/test/unit/bookmark_editor_test.js [hunk 7]
```diff
@@ -13,7 +15,8 @@ require('/shared/js/url_helper.js');
 
 suite('bookmark_editor.js >', function() {
 
-  var getStub, iconRenderStub;
+  var getStub, iconRenderStub, realL10n,
+  RealWebManifestHelper;
```
Comment (crdlc) [inline apps/bookmark/test/unit/bookmark_editor_test.js]: s/RealWebManifestHelper/realWebManifestHelper better?

File: apps/bookmark/test/unit/mock_web_manifest_helper.js [hunk 8]
```diff
@@ -0,0 +1,10 @@
+'use strict';
+
+/* exported MockWebManifestHelper */
+var MockWebManifestHelper = {
+  getManifest: function(manifestURL) {
+    return new Promise(function(resolve, reject) {
+      resolve({'short_name': 'App', 'name': 'My App'});
+    });
+  }
+};
```
Comment (crdlc) [inline apps/bookmark/test/unit/mock_web_manifest_helper.js]: new line

File: apps/bookmark/test/unit/mock_web_manifest_helper.js [hunk 9]
```diff
@@ -0,0 +1,10 @@
+'use strict';
+
+/* exported MockWebManifestHelper */
+var MockWebManifestHelper = {
+  getManifest: function(manifestURL) {
+    return new Promise(function(resolve, reject) {
```
Comment (crdlc) [inline apps/bookmark/test/unit/mock_web_manifest_helper.js]:
```js
return Promise.resolve({'short_name': 'App', 'name': 'My App'});
```

File: apps/bookmark/test/unit/mock_xml_http_request.js [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+  this.onload();
+};
+
+MockXMLHttpRequest.prototype.setRequestHeader = function() {
+
+};
+
+MockXMLHttpRequest.prototype.open = function() {
+
+};
```
Comment (crdlc) [inline apps/bookmark/test/unit/mock_xml_http_request.js]: new line

File: apps/bookmark/test/unit/web_manifest_helper_test.js [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+  });
+
+  teardown(function() {
+    window.XMLHttpRequest = RealXMLHttpRequest;
+  });
+
+  test('getManifest()', function(done) {
+    WebManifestHelper.getManifest().then(
+      function (result) {
+          assert.equal(result, 'foo');
```
Comment (crdlc) [inline apps/bookmark/test/unit/web_manifest_helper_test.js]: wrong indentation, 2 spaces for this line

File: apps/bookmark/test/unit/web_manifest_helper_test.js [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+          done();
```
Comment (crdlc) [inline apps/bookmark/test/unit/web_manifest_helper_test.js]: same here

File: apps/bookmark/js/web_manifest_helper.js [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+        var status = xhr.status;
+        if (status == 200) {
+          resolve(xhr.response);
+        } else {
+          reject(status);
+        }
+      };
+      xhr.onerror = function(e) {
+        console.error('Unable to get web manifest');
+        reject();
```
Comment (crdlc) [inline apps/bookmark/js/web_manifest_helper.js]: reject(e.target.status) ?

File: apps/bookmark/js/bookmark_editor.js (hunk 3 above)
Comment (benfrancis) [inline apps/bookmark/js/bookmark_editor.js]: Yep, OK

File: apps/bookmark/js/bookmark_editor.js (hunk 4 above)
Comment (benfrancis) [inline apps/bookmark/js/bookmark_editor.js]: Hmm, I was using "get" in the sense of HTTP GET, but I can see how that could be confusing. I changed it to "_fetchManifest"

File: apps/bookmark/js/bookmark_editor.js (hunk 2 above)
Comment (benfrancis) [inline apps/bookmark/js/bookmark_editor.js]: Oh yes, I think in a previous version I needed to store the manifest data on the BookmarkEditor object, but that's redundant now.

File: apps/bookmark/js/web_manifest_helper.js (hunk 13 above)
Comment (benfrancis) [inline apps/bookmark/js/web_manifest_helper.js]: yep

File: apps/bookmark/test/unit/bookmark_editor_test.js (hunk 7 above)
Comment (benfrancis) [inline apps/bookmark/test/unit/bookmark_editor_test.js]: yep

**Comment by try-server-hook** [comment]
[benfrancis Ben Francis (benfrancis)](https://github.com/benfrancis) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=b98d75d1d7b3)

**Comment by try-server-hook** [comment]
[benfrancis Ben Francis (benfrancis)](https://github.com/benfrancis) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=8f7dc4aff6ff)

File: apps/bookmark/js/bookmark_editor.js [hunk 14]
```diff
@@ -87,9 +119,14 @@ var BookmarkEditor = {
...
     this.saveButton.disabled = title === '';
   },
+  
+  _installApp: function bookmarkEditor_installApp() {
+    window.navigator.mozApps.install(this.manifestURL);
+  },
 
   save: function bookmarkEditor_save(evt) {
     this.saveButton.removeEventListener('click', this.saveListener);
+    this.installAppButton.removeEventListener('click', this._installApp);
```
Comment (crdlc) [inline apps/bookmark/js/bookmark_editor.js]: it was added using binding, please save the reference to this._installApp.bind(this) to remove this listener, thanks a lot

**Comment by try-server-hook** [comment]
[benfrancis Ben Francis (benfrancis)](https://github.com/benfrancis) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=0f9a7ee98f5b)

**Pull request merged**